### PR TITLE
upkeep/e2e-utils-playwright: Exposing Playwright utils

### PR DESCRIPTION
**Pull request by Sidsector9**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR exposes the Playwright E2E utils as a consumable NPM package so that extension developers can take advantage of common flows such as Checkout, Cart, etc.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

Expose Playwright E2E tests utils.

</details>

### CODE REVIEW

**Comment by Sidsector9** [comment]
@adimoldovan tagging you to confirm the approach taken in this PR.

**Comment by adimoldovan** [comment]
> @adimoldovan tagging you to confirm the approach taken in this PR.

Thanks @Sidsector9 for working on this. 
On a first check, I would suggest the following:
- start with only a few utils instead of adding all of them, and we'll iterate and add utils as needed. I'm saying this because many of those functions are not used in core and we'll be better of removing them. This would be a good chance to do that.
- decouple the test data. I don't think it should be in here.

**Comment by Sidsector9** [comment]
Thanks for checking it out @adimoldovan 

> decouple the test data. I don't think it should be in here.

Should the test data be a separate package? If not, can you point me where it should live?

**Comment by adimoldovan** [comment]
> Thanks for checking it out @adimoldovan
> 
> > decouple the test data. I don't think it should be in here.
> 
> Should the test data be a separate package? If not, can you point me where it should live?

I think it should live with the tests, in the each e2e project using the utils package.

Unless I missed something, the only thing required from the test-data is the admin credentials to authenticate an API request. That function should have the credentials as arguments instead.

**Comment by rodelgc** [comment]
Hi @Sidsector9 this pull request also has some merge conflicts. If you could also resolve them on your next commit, that would be great. Thanks!

**Comment by Sidsector9** [comment]
@rodelgc Thanks for the ping, I've updated the PR.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+export const closeChoosePatternModal = async ( { page } ) => {
+	const closeModal = page
+		.getByLabel( 'Scrollable section' )
+		.filter()
+		.getByRole( 'button', {
+			name: 'Close',
+			exact: true,
+		} );
+	await page.addLocatorHandler( closeModal, async () => {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js:9]: I don't think this will work as it's a custom method and isn't defined in the package.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	const responsePromise = page.waitForResponse(
+		( response ) =>
+			response.url().includes( '//single' ) && response.status() === 200
+	);
+	await page.goto( 'wp-admin/post-new.php' );
+	await disableWelcomeModal( { page } );
+	await responsePromise;
+};
+
+export const fillPageTitle = async ( page, title ) => {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js]: I'm not crazy about this utility.  It's not a real savings and just obfuscates what's already simple built-in methods.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	} else {
+		await page.getByLabel( 'Settings', { exact: true } ).click();
+	}
+};
+
+export const getCanvas = async ( page ) => {
+	return page.frame( 'editor-canvas' ) || page;
+};
+
+export const goToPageEditor = async ( { page } ) => {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js:42]: This seems like overkill just to load up a new page editor.  I think we can probably eliminate this and simplify things.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		await closeModal.click();
+	} );
+};
+
+export const disableWelcomeModal = async ( { page } ) => {
+	// Further info: https://github.com/woocommerce/woocommerce/pull/45856/
+	await page.waitForLoadState( 'domcontentloaded' );
+
+	const isWelcomeGuideActive = await page.evaluate( () =>
+		wp.data.select( 'core/edit-post' ).isFeatureActive( 'welcomeGuide' )
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js]: This is relying on a plugin that's installed during wp-env setup, but only in our e2e testing setup.  I worry that putting this into a utility will lead to folks having trouble with this if they're testing against other environments.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+		( response ) =>
+			response.url().includes( '//page' ) && response.status() === 200
+	);
+	await page.goto( 'wp-admin/post-new.php?post_type=page' );
+	await disableWelcomeModal( { page } );
+	await closeChoosePatternModal( { page } );
+	await responsePromise;
+};
+
+export const goToPostEditor = async ( { page } ) => {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js:48]: Same.  Why so much code just to load the post editor?

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+		} )
+		.click();
+	await page.getByPlaceholder( 'Search', { exact: true } ).fill( blockName );
+	await page.getByRole( 'option', { name: blockName, exact: true } ).click();
+
+	// In WP 6.6 'Toggle block inserter' button closes the inserter as expected,
+	// but trying to immediately open it again will fail in Playwright, while manually it works.
+	// We have tests that insert multiple blocks and fail because of this.
+	// Using the new 'Close block inserter' button added in WP 6.6 works fine.
+	if ( wpVersion && wpVersion <= 6.5 ) {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js]: I'm hesitant to put WP-version-specific code into a utilities package.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+	await canvas
+		.getByRole( 'document', {
+			name: 'Empty block; start writing or type forward slash to choose a block',
+		} )
+		.pressSequentially( `/${ blockName }` );
+	await expect(
+		page.getByRole( 'option', { name: blockName, exact: true } )
+	).toBeVisible();
+	await page.getByRole( 'option', { name: blockName, exact: true } ).click();
+	await expect(
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js]: We probably shouldn't be putting assertions into utilities.  We should remove these and instead have the assertion separate.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+		canvas.locator(
+			'.wp-block-woocommerce-classic-shortcode__placeholder-copy'
+		)
+	).toBeVisible();
+	await canvas
+		.getByRole( 'button' )
+		.filter( { hasText: 'Transform into blocks' } )
+		.click();
+
+	await expect( page.getByLabel( 'Dismiss this notice' ) ).toContainText(
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js]: Same, should remove the assertion from the utility.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	await expect(
+		canvas.getByLabel( `Block: ${ blockName }` ).first()
+	).toBeVisible();
+};
+
+export const transformIntoBlocks = async ( page ) => {
+	const canvas = await getCanvas( page );
+
+	await expect(
+		canvas.locator(
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/editor.js]: Same, should remove the assertion.

File: packages/js/e2e-utils-playwright/src/options.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Internal dependencies
+ */
+import { encodeCredentials } from './plugin-utils';
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/options.js]: Again, this relies on a plugin that's installed and activated during the specific wp-env setup.  I think we should avoid having this package have a dependency on an external plugin that may or may not be present in the test environment.

File: packages/js/e2e-utils-playwright/src/pdp.js [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+export async function addProductsToCart( page, productName, quantityCount ) {
+	await page.goto(
+		`product/${ productName.replace( / /gi, '-' ).toLowerCase() }`
+	);
+	await expect(
+		await page.getByRole( 'heading', { name: productName } ).count()
+	).toBeGreaterThan( 0 );
+	await page.getByLabel( 'Product quantity' ).fill( quantityCount );
+	await page.locator( 'button[name="add-to-cart"]' ).click();
+	await expect(
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/pdp.js]: We should remove the assertion.

File: packages/js/e2e-utils-playwright/src/pdp.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
+const { expect } = require( '@playwright/test' );
+
+/**
+ * Util helper made for adding multiple same products to cart
+ *
+ * @param page
+ * @param productName
+ * @param quantityCount
+ */
+export async function addProductsToCart( page, productName, quantityCount ) {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/pdp.js]: I'm not crazy about the name of this function.  `addOneOrMoreProduct()` might be better?  That way it can't be confused that it adds different products.

File: packages/js/e2e-utils-playwright/src/pdp.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+	await page.goto(
+		`product/${ productName.replace( / /gi, '-' ).toLowerCase() }`
+	);
+	await expect(
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/pdp.js]: Remove assertion.

File: packages/js/e2e-utils-playwright/src/plugin-utils.js [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+ *
+ * @param {string} url The URL where the zip file is located. Takes precedence over `repository`.
+ * @param {string} repository The repository owner and name. For example: `woocommerce/woocommerce`. Ignored when `url` was given.
+ * @param {string} authorizationToken Authorization token used to authenticate with the GitHub API if required.
+ * @param {boolean} prerelease Flag on whether to get a prelease or not. Default `false`.
+ * @param {string} downloadDir Relative path to the download directory. Non-existing folders will be auto-created. Defaults to `tmp` under current working directory.
+ *
+ * @return {string} Absolute path to the downloaded zip.
+ */
+export const downloadZip = async ( {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/plugin-utils.js]: I don't know that this utility is in use anywhere anymore.  It might be leftover from days when we were installing updated versions of WC on remote web hosts.

Comment (lanej0) [reply packages/js/e2e-utils-playwright/src/plugin-utils.js]: Regardless, I don't think it belongs in a general "Playwright test utilities" package.  I think we can remove it altogether.

File: packages/js/e2e-utils-playwright/src/simple-products.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+		} )
+		.click();
+}
+
+/**
+ * This function simulates the clicking of the "Add New" link under the "product" section in the menu.
+ *
+ * @param {import('@playwright/test').Page} page
+ */
+export async function clickAddNewMenuItem( page ) {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/simple-products.js]: We really don't need a helper utility to click a link.  Playwright has a built in function for that.

File: packages/js/e2e-utils-playwright/src/simple-products.js [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+		.getByRole( 'link', { name: 'Add New' } )
+		.click();
+}
+
+/**
+ * This function checks if the old product editor is visible.
+ *
+ * @param {import('@playwright/test').Page} page
+ */
+export async function expectOldProductEditor( page ) {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/simple-products.js]: Again, this has very little utility.  It should just be an assertion within the test.

File: packages/js/e2e-utils-playwright/src/simple-products.js [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+		page.getByRole( 'heading', { name: 'Product data' } )
+	).toBeVisible();
+}
+
+/**
+ * This function checks if the block product editor is visible.
+ *
+ * @param {import('@playwright/test').Page} page
+ */
+export async function expectBlockProductEditor( page ) {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/simple-products.js]: Same as above.

File: packages/js/e2e-utils-playwright/src/simple-products.js [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+	).toContainText( 'Add new product' );
+}
+
+/**
+ * Click on a block editor tab.
+ *
+ * @param {string}                          tabName
+ * @param {import('@playwright/test').Page} page
+ */
+export async function clickOnTab( tabName, page ) {
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/simple-products.js]: Same again.

File: packages/js/e2e-utils-playwright/src/themes.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
+const { request } = require( '@playwright/test' );
+
+import { encodeCredentials } from './plugin-utils';
```
Comment (lanej0) [inline packages/js/e2e-utils-playwright/src/themes.js]: Also relies on that extra plugin, so we should not include it in a general utilities package.

File: tools/storybook/package.json [hunk 20]
```diff
@@ -109,6 +109,9 @@
 				"node_modules/@woocommerce/eslint-plugin/configs",
 				"node_modules/@woocommerce/eslint-plugin/rules",
 				"node_modules/@woocommerce/eslint-plugin/index.js",
+				"node_modules/@woocommerce/block-library/build",
+				"node_modules/@woocommerce/block-library/blocks.ini",
+				"node_modules/@woocommerce/admin-library/build",
```
Comment (lanej0) [inline tools/storybook/package.json]: Is this change intentional?  Not sure why we're changing files in storybook?

**Review by lanej0 — CHANGES_REQUESTED** [review]
While I agree in principle that it's useful to create a utilities package, I don't like the idea of copying over the helpers that are currently scattered around the various e2e tests.  I think we need to be really intentional with what is included in the utilities package, and reject utilities with low value or that introduce additional dependencies.  

I think this is a good start, but we should scale the initial implementation way back and really think about what should be implemented.  My opinion is that it should contain:

- The add product utility is good
- The order number helper is useful
- Some of the blocks editing utilities could be helpful, but they need a re-think/refactor
- The checkout and cart utilities are a good start

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 3 above)
Comment (Sidsector9) [reply packages/js/e2e-utils-playwright/src/editor.js:42]: @lanej0 I've migrated all the utils that are actively used in the tests. I have simplified it/

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 6 above)
Comment (Sidsector9) [reply packages/js/e2e-utils-playwright/src/editor.js]: @lanej0 how would you suggest we handle this?

File: packages/js/e2e-utils-playwright/src/simple-products.js (hunk 18 above)
Comment (Sidsector9) [reply packages/js/e2e-utils-playwright/src/simple-products.js]: @lanej0 are you sure to remove this? This is actively used in a lot of tests.

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 6 above)
Comment (lanej0) [reply packages/js/e2e-utils-playwright/src/editor.js]: I'd say leave it out of the package.  It can continue to be used as a helper function and eventually deprecated.

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 3 above)
Comment (lanej0) [reply packages/js/e2e-utils-playwright/src/editor.js:42]: Okay, not hurting anything, so we'll leave it as is for now!

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 1 above)
Comment (lanej0) [reply packages/js/e2e-utils-playwright/src/editor.js:9]: @Sidsector9 Did you confirm that this works?  `addLocatorHandler` is a custom method, and I didn't see it defined anywhere in the package.

**Review by lanej0 — APPROVED** [review]
This looks like a good start to this.  I really appreciate all of the work you've put in here!  I had one question about the `addLocatorHandler` method -- but otherwise everything else looks good. 👍🏼 from me!

**Comment by lanej0** [comment]
Looks like we'll just need a changelog and a quick resolution to the pnpm-lock conflict and this should be good to go!

**Comment by adimoldovan** [comment]
I see some (apparently) unrelated unit tests failing that should be fixed as well. The only jobs that are ok to fail are the `CI / Test reports` ones and the `Build Live Branch`, and those only because this is a fork and cannot access the GH token.
Not sure about why the unit tests are failing, but it could be related to the pnpm version. @Sidsector9 please make sure you're using pnpm 9.1.3 before recreating the pnpm lock file.

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 1 above)
Comment (adimoldovan) [reply packages/js/e2e-utils-playwright/src/editor.js:9]: @lanej0 `addLocatorHandler` is a builtin Playwright method. We're good. https://playwright.dev/docs/api/class-page#page-add-locator-handler

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 3 above)
Comment (adimoldovan) [reply packages/js/e2e-utils-playwright/src/editor.js:42]: I think this is acceptable as a util, because it also disabled/closes 2 modals - the reason it exists as a method. I would however rename it to be explicit on its function. Something like `goToPageEditorAndDisableModals` would be better I think.

**Comment by Sidsector9** [comment]
@adimoldovan this is my setup:
- Node: v20.17.0
- PNPM: 9.1.3

I've rebuilt the lock file, but unable to find the reason from the logs as to why those tests are failing.

**Comment by adimoldovan** [comment]
> @adimoldovan this is my setup:
> 
> * Node: v20.17.0
> * PNPM: 9.1.3
> 
> I've rebuilt the lock file, but unable to find the reason from the logs as to why those tests are failing.

Thanks for checking @Sidsector9. pnpm >9.1.3 used to cause some trouble, seems like this is not the case here. I cannot say out what's the issue here from a quick view, but I'm pretty sure the pnpm lock change causes the failures.

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 21]
```diff
@@ -0,0 +1,121 @@
+import { select, dispatch } from '@wordpress/data';
```
Comment (adimoldovan) [inline packages/js/e2e-utils-playwright/src/editor.js]: You don't need this import.

File: packages/js/e2e-utils-playwright/package.json [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+	},
+	"license": "GPL-3.0+",
+	"engines": {
+		"node": "^20.11.1",
+		"pnpm": "9.1.3"
+	},
+	"main": "src/index.js",
+	"module": "build-module/index.js",
+	"dependencies": {
+		"@wordpress/data": "wp-6.0"
```
Comment (adimoldovan) [inline packages/js/e2e-utils-playwright/package.json]: This package is not used in any of the utils. The import in editor.js is not needed.

**Review by adimoldovan — CHANGES_REQUESTED** [review]

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+		await closeModal.click();
+	} );
+};
+
+export const disableWelcomeModal = async ( { page } ) => {
+	// Further info: https://github.com/woocommerce/woocommerce/pull/45856/
+	await page.waitForLoadState( 'domcontentloaded' );
+
+	const isWelcomeGuideActive = await page.evaluate( () =>
+		select( 'core/edit-post' ).isFeatureActive( 'welcomeGuide' )
```
Comment (adimoldovan) [inline packages/js/e2e-utils-playwright/src/editor.js]:
```suggestion
		window.wp.data.select( 'core/edit-post' ).isFeatureActive( 'welcomeGuide' )
```

File: packages/js/e2e-utils-playwright/src/editor.js [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+	);
+
+	if ( isWelcomeGuideActive ) {
+		await page.evaluate( () =>
+			dispatch( 'core/edit-post' ).toggleFeature( 'welcomeGuide' )
```
Comment (adimoldovan) [inline packages/js/e2e-utils-playwright/src/editor.js]:
```suggestion
			window.wp.data.dispatch( 'core/edit-post' ).toggleFeature( 'welcomeGuide' )
```

File: packages/js/e2e-utils-playwright/src/editor.js (hunk 23 above)
Comment (Sidsector9) [reply packages/js/e2e-utils-playwright/src/editor.js]: @adimoldovan this was added after this review: https://github.com/woocommerce/woocommerce/pull/51982#discussion_r1815717274

**Comment by adimoldovan** [comment]
@Sidsector9 Thanks for the updates. I see the failures were resolved by removing the extra dependency.

A changelog is still required, so that `CI / Validate changelog` check passes.
You can add a changelog in 2 ways:
- Update the "Changelog entry section" in PR's description and a changelog will be automatically generated by a Github action.
- Generate it locally by running `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and following the instructions and commit the file.

**Review by adimoldovan — APPROVED** [review]
@Sidsector9 This looks good. Let's merge it.

**Comment by adimoldovan** [comment]
Tracking follow-up tasks in https://github.com/woocommerce/woocommerce/issues/52939 epic.
